overlord/devicestate/devicestate.go

```diff
@@ -738,66 +742,70 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
-			ts, err = remodelVar.InstallWithDeviceContext(ctx, st,
+			ts, err := remodelVar.InstallWithDeviceContext(ctx, st,
 				snapPathSi, modelSnap.SnapName(), newModelSnapChannel,
 				userID, snapstate.Flags{Required: true}, deviceCtx,
 				fromChange)
 			if err != nil {
 				return nil, err
 			}
 			tss = append(tss, ts)
-		} else if currentInfo != nil && newModelSnapChannel != "" {
-			// the snap is already installed and has its default
-			// channel declared in the model, but the local install
-			// may be tracking a different channel
-			changed, err := installedSnapChannelChanged(st, modelSnap.SnapName(), newModelSnapChannel)
-			if err != nil {
+
+			if err := updateNeededSnapsFromTs(ts); err != nil {
 				return nil, err
 			}
-			if changed {
-				ts, err = remodelVar.UpdateWithDeviceContext(st,
-					snapPathSi, modelSnap.SnapName(),
-					newModelSnapChannel, userID,
-					snapstate.Flags{NoReRefresh: true},
-					deviceCtx, fromChange)
-				if err != nil {
-					return nil, err
-				}
-				tss = append(tss, ts)
-			}
+
+			continue
 		}
-		if currentInfo == nil {
-			// snap is not installed, we have a task set then
-			if err := updateNeededSnapsFromTs(ts); err != nil {
+
+		// the snap is already installed and has its default channel declared in
+		// the model, but the local install may be tracking a different channel
+		changed, err := installedSnapChannelChanged(st, modelSnap.SnapName(), newModelSnapChannel)
+		if err != nil {
+			return nil, err
+		}
+
+		// snap is installed already, so we have 2 possible scenarios:
+		// 1. the snap will be updated (new channel or revision), in which case
+		// we should make sure that the prerequisites of the new revision are
+		// accounted for
+		// 2. the snap channel or revision is not being modified so grab
+		// whatever is required for the current revision
+		if changed {
+			ts, err := remodelVar.UpdateWithDeviceContext(st,
+				snapPathSi, modelSnap.SnapName(),
+				newModelSnapChannel, userID,
+				snapstate.Flags{NoReRefresh: true},
+				deviceCtx, fromChange)
+			if err != nil {
 				return nil, err
 			}
-		} else {
-			// snap is installed already, so we have 2 possible
-			// scenarios, one the snap will be updated, in which
-			// case we have a task set and should make sure that the
-			// prerequisites of the new revision are accounted for,
-			// or two, the snap revision is not being modified so
-			// grab whatever is required for the current revision
-			if ts != nil && ts.MaybeEdge(snapstate.LastBeforeLocalModificationsEdge) != nil {
-				// not a simple task snap-switch-channel, so
-				// take the prerequisites needed by the new
-				// revision we're updating to
+			tss = append(tss, ts)
+
+			// if this is true, then the snap's revision was changed, and we
+			// need to extract the prerequisites from the task set. otherwise,
+			// we treat the snap as if it were unchanged.
+			if ts.MaybeEdge(snapstate.LastBeforeLocalModificationsEdge) != nil {
```

Comment:
So, iiuc basically we know that the revision has not changed because no edge separating tasks that do not modify the system from those modifying the system state has been set. Which is set only if the latter tasks exist, and if they don't we are not really refreshing to a different to the current one revision. This is fine and is what we were doing already, but seems very indirect... I wonder if there would be a more explicit way. Anyway, it would be good if you can expand a bit the comment if what I wrote is correct.

Reply:
I agree that the expression doesn't clearly indicate the true meaning. Maybe I could replace it with a little function called `didRevisionChangeFromTS` or something like that, for clarity? Either way, I'll add the comment. I also think an explanation of the general concept of edges between tasks would be helpful for my understanding.

Reply:
Comment updated, let me know if you think that makes it a bit more clear